File: apps/web/src/lib/api/storage.ts
import { apiFetch } from "./core"

export type StorageProvider = "s3" | "r2" | "minio" | "b2"

export interface ApiStorageIntegration {
  id: string
  organization_id: string
  name: string
  provider: StorageProvider
  endpoint: string
  region: string
  bucket: string
  created_at: string
  updated_at: string
}

export interface CreateStorageBody {
  name: string
  provider: StorageProvider
  endpoint?: string
  region?: string
  bucket: string
  access_key_id: string
  secret_access_key: string
}

export const storage = {
  list: (orgId: string, token: string) =>
    apiFetch<ApiStorageIntegration[]>(
      `/api/v1/orgs/${orgId}/storage-integrations`,
      {},
      token
    ),

  create: (orgId: string, body: CreateStorageBody, token: string) =>
    apiFetch<ApiStorageIntegration>(
      `/api/v1/orgs/${orgId}/storage-integrations`,
      { method: "POST", body: JSON.stringify(body) },
      token
    ),

  delete: (orgId: string, id: string, token: string) =>
    apiFetch<void>(
      `/api/v1/orgs/${orgId}/storage-integrations/${id}`,
      { method: "DELETE" },
      token
    ),
}

File: packages/server/service/storage.go
package service

import (
	"context"

	"github.com/google/uuid"
	"github.com/meshploy/packages/db"
	"gorm.io/gorm"
)

type StorageService struct {
	db *gorm.DB
}

type CreateStorageInput struct {
	Name            string
	Provider        db.StorageProvider
	Endpoint        string
	Region          string
	Bucket          string
	AccessKeyID     string
	SecretAccessKey string
}

func (s *StorageService) List(ctx context.Context, orgID uuid.UUID) ([]db.StorageIntegration, error) {
	items := make([]db.StorageIntegration, 0)
	err := s.db.WithContext(ctx).Where("organization_id = ?", orgID).Find(&items).Error
	return items, err
}

func (s *StorageService) Create(ctx context.Context, orgID uuid.UUID, in CreateStorageInput) (*db.StorageIntegration, error) {
	item := &db.StorageIntegration{
		OrganizationID:  orgID,
		Name:            in.Name,
		Provider:        in.Provider,
		Endpoint:        in.Endpoint,
		Region:          in.Region,
		Bucket:          in.Bucket,
		AccessKeyID:     db.EncryptedString(in.AccessKeyID),
		SecretAccessKey: db.EncryptedString(in.SecretAccessKey),
	}
	return item, s.db.WithContext(ctx).Create(item).Error
}

func (s *StorageService) Delete(ctx context.Context, id, orgID uuid.UUID) error {
	return s.db.WithContext(ctx).
		Where("id = ? AND organization_id = ?", id, orgID).
		Delete(&db.StorageIntegration{}).Error
}
